contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,189 @@
+
+    /**
+     * @dev The `sides` and `zero` arrays are set, at initialization, to have a length equal to the depth of the tree.
+     * No push/pop operations should be performed of these array, and their lengths should not be updated.
+     *
+     * The `roots` array stores the history of roots. Its length is set at initialization, and should not be updated.
+     *
+     * The hashing function used during initialization to compute the `zeros` values (value of a node at a given depth
+     * for which the subtree is full of zero leaves). This function is kept in the structure for handling insertions.
+     */
+    struct TreeWithHistory {
+        uint256 currentRootIndex;
+        uint256 nextLeafIndex;
+        bytes32[] sides;
+        bytes32[] zeros;
+        bytes32[] roots;
+        function(bytes32, bytes32) view returns (bytes32) fnHash;
+    }
+
+    /**
+     * @dev Initialize using the default hash
+     */
+    function initialize(TreeWithHistory storage self, uint256 depth, uint256 length, bytes32 zero) internal {
+        return initialize(self, depth, length, zero, MerkleProof.hashPair);
+    }
+
+    /**
+     * @dev Initialize a new complete MerkleTree defined by:
+     * - Depth `depth`
+     * - All leaves are initialize to `zero`
+     * - Hashing function for a pair of leaves is fnHash
+     * and keep a root history of length `length` when leaves are inserted.
+     */
+    function initialize(
+        TreeWithHistory storage self,
+        uint256 depth,
+        uint256 length,
+        bytes32 zero,
+        function(bytes32, bytes32) view returns (bytes32) fnHash
+    ) internal {
+        if (depth > MAX_DEPTH) {
+            revert MerkleTreeInvalidDepth(depth, MAX_DEPTH);
+        }
+
+        // Store depth & length in the dynamic array
+        Arrays.unsafeSetLength(self.sides, depth);
+        Arrays.unsafeSetLength(self.zeros, depth);
+        Arrays.unsafeSetLength(self.roots, length);
+        self.fnHash = fnHash;
+
+        // Build the different hashes in a zero-filled complete tree
+        bytes32 currentZero = zero;
+        for (uint32 i = 0; i < depth; ++i) {
+            Arrays.unsafeAccess(self.zeros, i).value = currentZero;
+            currentZero = fnHash(currentZero, currentZero);
+        }
+
+        // Insert the first root
+        Arrays.unsafeAccess(self.roots, 0).value = currentZero;
+    }
+
+    /**
+     * @dev Insert a new leaf in the tree, compute the new root, and store that new root in the history.
+     */
+    function insert(TreeWithHistory storage self, bytes32 leaf) internal returns (uint256) {
+        // Cache read
+        uint256 depth = self.zeros.length;
+        function(bytes32, bytes32) view returns (bytes32) fnHash = self.fnHash;
+
+        // Get leaf index
+        uint256 leafIndex = self.nextLeafIndex++;
+
+        // Check if tree is full.
+        if (leafIndex == 1 << depth) Panic.panic(Panic.RESOURCE_ERROR);
```

Comment:
Shouldn't this ideally be?

```suggestion
        if (leafIndex >= 1 << depth) Panic.panic(Panic.RESOURCE_ERROR);
```

Reply:
unless someone starts manually tampering with leafIndex, it will only be increment one-by-one ... so it would have to be equal (and revert) before it is greater. 

But the cost of that change is not much, so if you feel like its safer ... I don't see an issue with this change.